### PR TITLE
Fix uvcal add

### LINKED ISSUE #1107: Fix `__add__` method in UVCal in the case of overlapping and differently sorted data
see #1102 and #1105 for reference

### PR DESCRIPTION
**Pull request by bhazelton**
## Description

Fix bugs in `UVCal.__add__` when data are sorted differently along any axis or interleaved along an axis. Inspired by #1105.

## Motivation and Context

Fixes #1107

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

Bug fix checklist:
- [x] My fix includes a new test that breaks as a result of the bug (if possible).
- [x] All new and existing tests pass.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/CHANGELOG.md).

### CODE REVIEW

File: pyuvdata/uvdata/uvdata.py
Comment (bhazelton) [inline pyuvdata/uvdata/uvdata.py:7545]: This if statement isn't needed -- the block before is in the same if statement.

File: pyuvdata/uvcal/uvcal.py
Comment (kartographer) [inline pyuvdata/uvcal/uvcal.py:2353]: 👍

File: pyuvdata/uvcal/tests/test_uvcal.py
Comment (kartographer) [inline pyuvdata/uvcal/tests/test_uvcal.py:2031]: So I worry this test might not actually poke at the underlying issue that was discovered, which appeared when you had overlapping-but-differently-ordered data. As far as I can tell, it seems like all of these tests split the data in half, and reorder the halves before recombination. It would be good to have at least some subset of these have that overlap present.

**Review by kartographer — COMMENTED** [review]
Though I'm not a uvcal expert, I think the implemented solution looks good. I did have a comment about the _tests_ that were added -- I think they just require a small modification. Otherwise, this is (also) looking good.

File: pyuvdata/uvcal/tests/test_uvcal.py
Comment (bhazelton) [reply pyuvdata/uvcal/tests/test_uvcal.py:2031]: Hmm, I tried to structure this test in the same way as the test you wrote for #1105. I don't think that was related to actually overlapping data (as in data for exactly the same blt, freq, pol) or we would have correctly errored. I thought it was when you had an axis that was not involved in the sum (so had all the same values on the objects) that was differently ordered.

File: pyuvdata/uvcal/uvcal.py
Comment (bhazelton) [reply pyuvdata/uvcal/uvcal.py:3248]: yes, that makes it much better!

Comment (bhazelton) [reply pyuvdata/uvcal/uvcal.py]: I made the analogous change in UVData as well.

File: pyuvdata/uvcal/tests/test_uvcal.py
Comment (kartographer) [reply pyuvdata/uvcal/tests/test_uvcal.py:2031]: Ahh, yes -- sorry, I missed the fact that you've got the equivalent of a double loop here. Clearly been too long since I've written a test...

File: pyuvdata/uvcal/uvcal.py
Comment (kartographer) [inline pyuvdata/uvcal/uvcal.py:3281]: Always good to see commits where the subtractions are twice the number of lines of additions 😉

**Review by kartographer — DISMISSED** [review]
Looks good! I've nominally approved the PR, but I'll leave it to you to merge in case you want someone else to take a look.

**Comment by bhazelton** [comment]
@kartographer I had to rebase after I merged the other PR because of conflicts in the changelog. Can you please re-approve?

**Review by kartographer — APPROVED** [review]

**Pull request merged**
